### PR TITLE
jobs

### LINKED ISSUE #33: Type/provider for managing jobs
We should have a basic way of managing the rundeck jobs using the api:

Here is a gist of roughly how it should work (this can be split into examples):

https://gist.github.com/liamjbennett/741c9139a66198dc38b2

### PR DESCRIPTION
**Pull request by jwbraucher**
#### Pull Request (PR) description
Adds support for installing the rundeck cli and using it to create rundeck jobs.

#### This Pull Request (PR) fixes the following issues
Fixes #33

### CODE REVIEW

**Comment by ghoneycutt** [comment]
@jwbraucher Thank you for your contribution! We appreciate new functionality though cannot merge without tests. Could you please add tests.

Check out https://github.com/voxpupuli/puppet-rundeck/blob/master/spec/defines/config/project_spec.rb as an example for a defined type.

File: manifests/config.pp [hunk 1]
```diff
@@ -99,6 +102,22 @@
     fail('when rundeck::manage_home = false a file definition for the home directory must be included outside of this module.')
   }
 
+  if $manage_cli_config {
+    file{ "${rdeck_home}/.rd":
+      ensure  => directory,
+    }
+    file{ "${rdeck_home}/.rd/rd.conf":
+      ensure  => file,
+      content => @("END")
```
Comment (ghoneycutt) [inline manifests/config.pp:111]: Suggest using a template instead of there HEREDOC as templates can be validated.

File: manifests/config/job.pp [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+define rundeck::config::job(
+  String $job_name = $title,
+  String $project,
+  String $job_definition = '',
+  String $format = 'yaml',
+  String $group = '-',
+  Enum['present', 'absent'] $ensure = 'present',
+) {
+
+  include rundeck::jobs
```
Comment (ghoneycutt) [inline manifests/config/job.pp:57]: should also `include rundeck` so that `$rundeck::rdeck_home` is available

File: manifests/config/job.pp [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+
+  if $ensure == 'present' {
+
+    # create job
+    file { "${group}-${job_name}":
+      ensure  => file,
+      path    => "${jobs_dir}/${job_filename}",
+      content => template($job_definition),
+    }
+    ~> exec { "${group}-${job_name}":
```
Comment (ghoneycutt) [inline manifests/config/job.pp:84]: this type of ordering is brittle, suggest having the file notify the exec.

File: manifests/jobs.pp [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+#
+
+class rundeck::jobs {
+
+  # pre-create directory to hold job definitions
+  $jobs_dir = "${rundeck::rdeck_home}/jobs"
+  ensure_resource(file, $jobs_dir, {'ensure' => 'directory'})
+
+  # create jobs
+  create_resources(rundeck::config::job, $rundeck::jobs)
```
Comment (ghoneycutt) [inline manifests/jobs.pp:14]: `create_resources()` is the old way of doing this, the newer way would be with `.each`

```puppet
  $rundeck::jobs.each |$name, $values| {
    rundeck::config::job { $name:
      * => $values,
    }
  }
```

**Comment by Joris29** [comment]
@kenyon I would like to add this cli option i was wondering about the approach and i would like your opinion on it.
Let me explain what i'm wondering about so rundeck and rundeck-cli both have different life-cycle and can be used separately.
So i'm wondering should we really include it in the rundeck class or have a complete separate class or even a separate module?

**Comment by kenyon** [comment]
@Joris29 I'd go with include in the rundeck class or a separate class, probably doesn't make sense to have a dedicated module, but I don't use rundeck so whatever makes most sense for you.

**Comment by Joris29** [comment]
Alright then I will go for a separate class for the cli and work something out

**Pull request closed without merging**